CONFIG.md
# Configuration
The application can be configured by setting the values in <a href="https://github.com/catchouli/better_tactics/blob/main/src/app.rs#L17">AppConfig</a>. This can be done either using environment variables, or a .env file. See <a href="https://github.com/catchouli/better_tactics/blob/main/.env">the development .env file</a> for an example. The application config is printed out on startup, and any errors parsing config values are logged as errors, so it's easy to check if your configuration is being picked up.

## Supported configuration values

# General
| Environment Variable | Default | Description |
| --- | --- | --- |
| BIND_INTERFACE | 127.0.0.1 | The IP address to bind the webserver to. By default, it binds to 127.0.0.1, so can only be accessed locally. To make it available externally, set BIND_INTERFACE to 0.0.0.0 to bind all interfaces. |
| BIND_PORT | 3030 | The port to bind the web server to, the default is 3030 |
| DATABASE_URL | sqlite://puzzles.sqlite | The url of the database. "sqlite://path/to/puzzles.sqlite" refers to the relative path "./path/to/puzzles.sqlite", while "sqlite:///path/to/puzzles.sqlite" with an additional slash at the start of the path refers to the absolute path /path/to/puzzles.sqlite |

# Tactics
| Environment Variable | Default | Description |
| --- | --- | --- |
| TACTICS_PUZZLE_RATING_VARIATION_DOWN | 0.05 | The percentage below your rating random puzzles can be shown on the 'next puzzle' page. 0.05 means 5%, so if the user's rating is 1000, the rating for selected puzzles can be as low as 950. |
| TACTICS_PUZZLE_RATING_VARIATION_UP | 0.0 | The percentage above your rating random puzzles can be shown on the 'next puzzle' page. This is set to 0% by default to avoid showing puzzles too high, but can be changed to the previous default of 0.05 to show puzzles a little above your current rating. |

# User Interface
| Environment Variable | Default | Description |
| --- | --- | --- |
| UI_INITIAL_MOVE_DELAY | 500 | The delay before the first computer move in ms when showing a puzzle |
| UI_SUBSEQUENT_MOVE_DELAY | 250 | The delay before the subsequent computer moves when in ms |

# Backups
Once enabled, backups will run once a day. If a backup is due to be run today and hasn't already, it will be attempted when the application starts. Automated backups are stored as .sqlite files with only the user data and no puzzles, but can be loaded back in as the main application database, at which point the puzzles will be automatically re-initialised from the lichess database.

| Environment Variable | Default | Description |
| --- | --- | --- |
| BACKUP_ENABLED | false | Whether or not daily automated backups are enabled |
| BACKUP_PATH | ./backups | The path to store automated backups |
| BACKUP_HOUR | 4 | The hour (local time) at which the automated backup is scheduled to run, i.e. 4 is 4am |

# Spaced repetition
See <a href="https://super-memory.com/english/ol/sm2.htm">Supermemo 2 algorithm</a> for information about the spaced repetition algorithm.

| Environment Variable | Default | Description |
| --- | --- | --- |
| SRS_DEFAULT_EASE | 2.5 | The default ease factor, the multiplier for card interval for each review, which goes up for 'easy' reviews and down for 'hard' or 'again' reviews |
| SRS_MINIMUM_EASE | 1.3 | The minimum ease factor when the ease is decreased |
| SRS_EASY_BONUS | 1.3 | The extra multiplier to the card interval when a card is review as 'easy' |
| SRS_DAY_END_HOUR | 4 | The hour (local time) at which the day is considered to start/end. The review queue will automatically include cards up to this time, so user can review all of today's cards at once. |
| SRS_REVIEW_ORDER | PuzzleRating | The order for puzzles to show up when reviewing. Valid values are: DueTime (the time the card is due), PuzzleRating (lower rated puzzles are shown first), and Random (reviews are shown in a random order from the pool of due reviews).|

## Deprecated configuration values
| Environment Variable | Description |
| ---  | --- |
| SQLITE_DB_NAME | Setting DATABASE_URL in the format "sqlite:///path/to/filename.sqlite" should be preferred. If DATABASE_URL is not set but SQLITE_DB_NAME is, DATABASE_URL will be initialised as "sqlite:///" + SQLITE_DB_NAME for compatibility. |


CHANGELOG.md
[1.11.0] - ????-??-??

Important: if you're building yourself, (i.e. not with the provided Dockerfile), npm is now needed
           in order to build the frontend. https://nodejs.org/en/download

Changed:
* The default rating variation on the 'Next puzzle' page is now 5% below the user's rating level,
  but not above the user's rating level anymore, as there was too much of a risk of showing puzzles
  that are too hard, especially as the user's rating level increased. There's now configuration
  values that can be changed to override this behavior if you preferred the previous +-5%.

[1.10.0] - 2023-11-02

Added:
* A hint button that shows a hint for the next move, for if the user can't work it out, but doesn't
  want to (or can't) go to lichess' analysis board to check the engine line. The puzzle will
  automatically be set to failed if the hint is used, but the user can always override it on the
  'Puzzle complete' panel and submit a positive response if desired.

Changed:
* Ctrl-c now causes a graceful shutdown and commits the database journal before closing it.
* The default interval for cards is now 1 minute. This doesn't affect much, but it makes it so that
  initially the 'Hard' and 'Good' buttons do something different (i.e. 1 minute and 10 minutes
  respectively), and so that pressing 'Again' on a puzzle makes it come up again in 1 minute
  instead of 10.

Fixed:
* A small fix to the scroll wheel behavior on the puzzle board to prevent the page from scrolling
  if the mousewheel is over the board.
* Fix countdown timer to next review not starting sometimes.
* The 'next review due' time on the stats screen no longer says 'never' when there are no reviews
  due for the rest of the day.

[1.9.0] - 2023-10-26

Added:
* Add simple move seeking (back and forward buttons) that allows you to see the moves played
  previously, and try playing out variations once the puzzle is over.
* Enable free movement after the wrong move is played the same as when the puzzle is complete,
  so you can play out variations and see what would have happened.
* Add config option to change the review order (SRS_REVIEW_ORDER) with three options:
  * DueTime (the current mode)
  * PuzzleRating (easier puzzles first)
  * Random (just show reviews in a random order)
* The default review order is now 'PuzzleRating', but the old behavior can be restored by using
  the option 'DueTime'.

Changed:
* Change initial computer move delay to 500, and make it and the subsequent move delays
  configurable with the configuration variables UI_INITIAL_MOVE_DELAY and UI_SUBSEQUENT_MOVE_DELAY.

[1.8.0] - 2023-10-21

Added:
* A Puzzle History page, which shows the history of all puzzles reviewed.
* On the 'next puzzle' screen, the puzzle to be reviewed is now saved until it is skipped.
* Skip buttons have been added to the puzzle interface for new puzzles:
  * "Skip this puzzle" when the puzzle is first started, which skips the current puzzle completely
    but doesn't award you any rating points.
  * "Too hard" and "Too easy" buttons, which also skip the puzzle, but correct your rating
    accordingly so you'll see easier or harder puzzles in the future.
  * "Don't repeat this puzzle", which marks the puzzle as complete, but without adding it to spaced
    repetition.
* Automated daily backups are now supported, and can be configured with the BACKUP_ENABLED,
  BACKUP_PATH, and BACKUP_HOUR configuration variables. See CONFIG.md for more details. The backup
  files are functional .sqlite databases that can be loaded back in as the application database,
  but do not include the puzzles, so if they are loaded the puzzles will be reinitialised when the
  application starts.

Removed:
* The configuration variable SQLITE_DB_NAME is now deprecated in favor of DATABASE_URL. If you
  have a SQLITE_DB_NAME set but not a DATABASE_URL, for compatibility DATABASE_URL will be set
  to "sqlite://{SQLITE_DB_NAME}". See CONFIG.md for more details.

Changed:
* Puzzles on the 'next puzzle' page will never be ones you've seen before anymore.
* Changes the .sqlite database to use WAL journaling (write-ahead-logging) so that multiple readers
  can use the database while something is writing to it. Overall the application should be a little
  faster.
* Tweaks the initial puzzle import to be a bit faster, and not lock the database as much. This is
  especially useful in docker, where the initial import was quite slow for me previously, and the
  web app was very slow to load while it was importing. The database import is now a lot faster,
  and page loads are generally instant while it is in progress. API calls that need to write (such
  as reviewing and skipping puzzles) might still need a couple of seconds while puzzles are being
  written to a database but it's much faster overall.

[1.7.0] - 2023-10-19

Changed:
* The puzzle page now loads the next puzzle automatically without refreshing the page.
* Cards leaving learning immediately due to reviewing them as 'Easy' now have their interval set
  to 4 days instead of the previous 1 day. This should allow actually easy puzzles to be seen far
  less frequently.
* The review forecast on the stats screen now allows you to change the number of days shown, and
  review scores allows you to see the data by percentage as well as total value.
* A link to the original lichess puzzle is now included on the puzzle interface.

[1.6.0] - 2023-10-13

Added:
* A review forecast on the stats page which shows the number of reviews due per day for the next
  week.
* A rating history which shows the user's rating growth over time.
* A review score histogram which shows the review scores being picked by the user by puzzle rating.

Changed:
* The user's rating deviation is now hidden from the stats page if it's under 100, the threshold
  under which the user's rating is no longer provisional and should be quite accurate.

[1.5.2] - 2023-10-09

Changed:

src/config.rs
use std::env::{self, VarError};
use std::error::Error;
use std::net::{IpAddr, Ipv4Addr};
use std::str::FromStr;

/// The application configuration.
#[derive(Debug, Clone)]
pub struct AppConfig {
    pub bind_interface: IpAddr,
    pub bind_port: u16,
    // TODO: change it to use the DATABASE_URL env var.
    pub db_name: String,
    pub srs: SrsConfig,
}

#[derive(Debug, Copy, Clone)]
pub struct SrsConfig {
    pub default_ease: f64,
    pub minimum_ease: f64,
    pub easy_bonus: f64,
}

impl Default for AppConfig {
    fn default() -> Self {
        Self {
            bind_interface: IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
            bind_port: 3030,
            db_name: "puzzles.sqlite".to_string(),
            srs: SrsConfig {
                default_ease: 2.5,
                minimum_ease: 1.3,
                easy_bonus: 1.3,
            },
        }
    }
}

impl AppConfig {
    /// Load the app config from a .env file or environment variables.
    pub fn from_env() -> Result<AppConfig, Box<dyn Error>> {
        let _ = dotenvy::dotenv();

        let defaults: AppConfig = Default::default();

        Ok(Self {
            bind_interface: Self::env_var_or_default("BIND_INTERFACE", defaults.bind_interface)?,
            bind_port: Self::env_var_or_default("BIND_PORT", defaults.bind_port)?,
            db_name: Self::env_var_or_default("SQLITE_DB_NAME", defaults.db_name)?,
            srs: SrsConfig {
                default_ease: Self::env_var_or_default("SRS_DEFAULT_EASE", defaults.srs.default_ease)?,
                minimum_ease: Self::env_var_or_default("SRS_MINIMUM_EASE", defaults.srs.minimum_ease)?,
                easy_bonus: Self::env_var_or_default("SRS_EASY_BONUS", defaults.srs.easy_bonus)?,
            },
        })
    }

    /// Read an env var and get a value, attempting to parse it to the specified type. If the
    /// variable is not set, returns the specified default.
    fn env_var_or_default<T>(key: &str, default: T) -> Result<T, Box<dyn Error>>
    where
        T: FromStr,
        <T as FromStr>::Err: Error + 'static,
    {
        match env::var(key) {
            Ok(value) => Ok(value.parse()?),
            Err(VarError::NotPresent) => Ok(default),
            Err(err) => Err(err.into())
        }
    }
}
